lib/fhir/common.js

```diff
@@ -119,6 +108,75 @@ module.exports = exports = (mongo) => {
     throw new Error(`Unknown ID generator '${generator}'. Valid options are 'uuidv1' or 'uuidv4'`)
   }
 
+  function resolveReference (reference, callback) {
+    const matches = REFERENCE_REGEX.exec(reference)
+    if (!matches) {
+      return callback(new Error(`Invalid resource reference "${reference}"`))
+    }
+    mongo.getDB((err, db) => {
+      if (err) {
+        return callback(err)
+      }
+      const resourceType = matches[4]
+      const resourceId = matches[5]
+      const versionId = matches[7]
+      if (versionId) {
+        db
+          .collection(`${resourceType}_history`)
+          .findOne({id: resourceId, 'meta.versionId': versionId}, callback)
+      } else {
+        db.collection(resourceType).findOne({id: resourceId}, callback)
+      }
+    })
+  }
+
+  const parseIncludeParams = (item, results) => {
+    return new Promise((resolve, reject) => {
+      let itemReference = item.split(':')[1]
+      resolveReference(itemReference, (err, data) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve(data)
+        }
+      })
+    })
+  }
+
+  function includeResources (ctx, results) {
+    return new Promise((resolve, reject) => {
+      if (!ctx) {
+        reject(`Invalid context parameter "${ctx}"`)
+      }
+      if (!Array.isArray(results)) {
+        reject(`Invalid results parameter "${results}"`)
+      }
+
+      let promises = []
+
+      if (ctx.query && ctx.query._include) {
+        if (Array.isArray(ctx.query._include)) {
+          ctx.query._include.forEach((item) => {
+            promises.push(parseIncludeParams(item, results))
+          })
+        } else {
+          promises.push(parseIncludeParams(ctx.query._include, results))
+        }
+      }
+
+      // wait for all references to be resolved before adding to results array
+      Promise.all(promises).then((res) => {
+        res.forEach((item) => {
+          results.push(item)
```

Comment:
It's not good practice to manipulate the arguments passed into a function. It would be better to use the [concat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat) array function here which does not change the existing array.

Reply:
👍